src/vm_cpu.rs

```diff
@@ -1102,63 +1098,69 @@ pub enum VmCpuStatus {
 struct VmCpusInner {
     // Locking: status must be locked before vcpu.
     status: RwLock<VmCpuStatus>,
-    vcpu: Mutex<VmCpu>,
+    vcpu: Once<PageBox<Mutex<VmCpu>>>,
+}
+
+impl Default for VmCpusInner {
+    fn default() -> Self {
+        VmCpusInner {
+            status: RwLock::new(VmCpuStatus::NotPresent),
+            vcpu: Once::new(),
+        }
+    }
 }
 
 /// The set of vCPUs in a VM.
 pub struct VmCpus {
-    inner: PageVec<VmCpusInner>,
+    num: RwLock<usize>,
+    vcpus: [VmCpusInner; MAX_CPUS],
 }
 
 impl VmCpus {
     /// Creates a new vCPU tracking structure backed by `pages`.
-    pub fn new(
-        guest_id: PageOwnerId,
-        pages: SequentialPages<InternalClean>,
-        page_tracker: PageTracker,
-    ) -> Result<Self> {
-        let num_vcpus = pages.length_bytes() / VM_CPU_BYTES;
-        if num_vcpus == 0 {
-            return Err(Error::InsufficientVmCpuStorage);
-        }
-        let mut inner = PageVec::new(pages, page_tracker);
-        for i in 0..num_vcpus {
-            let entry = VmCpusInner {
-                status: RwLock::new(VmCpuStatus::NotPresent),
-                vcpu: Mutex::new(VmCpu::new(i, guest_id)),
-            };
-            inner.push(entry);
+    pub fn new() -> Self {
+        let vcpus = [(); MAX_CPUS].map(|_| VmCpusInner::default());
+        Self {
+            num: RwLock::new(0),
+            vcpus,
         }
-        Ok(Self { inner })
     }
 
     /// Returns the number of vCPUs in this `VmCpus`.
     pub fn num_vcpus(&self) -> usize {
-        self.inner.len()
+        *self.num.read()
     }
 
     /// Adds the vCPU at `vcpu_id` as an available vCPU using `shared_area` as the vCPU's shared
     /// state-memory state area.
-    pub fn add_vcpu(&self, vcpu_id: u64, shared_area: VmCpuSharedArea) -> Result<()> {
-        let entry = self.inner.get(vcpu_id as usize).ok_or(Error::BadCpuId)?;
+    pub fn add_vcpu(&self, vcpu_id: u64, vcpu_box: PageBox<Mutex<VmCpu>>) -> Result<()> {
+        let entry = self.vcpus.get(vcpu_id as usize).ok_or(Error::BadCpuId)?;
         let mut status = entry.status.write();
         if *status != VmCpuStatus::NotPresent {
             return Err(Error::VmCpuExists);
         }
-        entry.vcpu.lock().shared_area.call_once(|| shared_area);
         *status = VmCpuStatus::PoweredOff;
+
+        assert!(!entry.vcpu.is_completed());
```

Comment:
why assert here instead of returning an error?

Reply:
Because this is an invariant failure. The VmCpuExists error above should cover the case where a vCPU has already been added. This is paranoia added because call_once will silently not do anything in case something is there already.

If this assert is triggered, we would have a vCPU that was marked as non present but somehow had some vCPU state already allocated. This shouldn't happen, and it would be disastrous if by mistake it would. Safe place to have an assert. :)

Reply:
Can that invariant be represented with the type system? A type that indicates existence and holds the object instead of an object and a separate flag?

Reply:
I thought about it. Essentially what we'd need is a guaranteed fixed-space vector (where `with_capacity` does in fact guarantee to allocate precisely that).
I didn't want to use third party crates for something like this, given the nature of the project.
The idea of adding yet another type, and the question of where to add it -- I think `vm_cpu.rs` is already too long than it should be -- pushed me to attempt a C-style approach, and make the type that .guarantees this `VmCpus` itself.

I am open to suggestion on how to do this more in line with where this projects needs to be.

Reply:
Something like this?

```
enum VmCpuState {
    NotPresent,
    PoweredOff(PageBox<Mutex<VmCpu>>),
    Runnable(PageBox<Mutex<VmCpu>>),
    Running(PageBox<Mutex<VmCpu>>),
}

struct VmCpuInner {
    vcpu: RwLock<VmCpuState>,
}
```

I think I tried this before the separate struct + flag approach that we have now and it was a little ugly. Maybe worth revisiting?

Reply:
Oh yes, I think I was missing the point by mixing up with the num issue below.